### ISSUE TITLE
Being cited for ShadowingOuterLocalVariable when the variable in question is in a different branch of a conditional statement

### ISSUE THREAD

**Issue by avtrujillo** [bbatsov/rubocop#4188]
## Expected behavior

I don't think this should be an offense, but maybe there are reasons that it should be. Figured I should report it regardless.

## Actual behavior

Rubocop gives me the following citation:

```
Test.rb:7:24: W: Shadowing outer local variable - animal.
  puts(animals.find { |animal| animal.length == 5})
                       ^^^^^^
```

when I ask it to inspect the following code:

```
# Test.rb

animals = ["dog", "cat", "chicken", "shark"]
 
if 2 + 2 == 5 # obviously this condition will never be true
  # at this point, animal is still undefined
  animal = "frog"
  puts animal # this line is only here so I don't get cited for useless assignment to animal
else
  # animal is now defined as nil, rather than
  # being undefined or defined as "frog"
  puts(animals.find { |animal| animal.length == 5})
  # Rubocop will cite me here for shadowing, becuase it considers animal to
  # still be defined.
end
```

## Steps to reproduce the problem

To see only the relevant offense, run the line below in a terminal:

```
rubocop Test.rb --only Lint
```

## RuboCop version

```
$ rubocop -V
0.48.0 (using Parser 2.4.0.0, running on ruby 2.3.3 x86_64-linux)

```

**Comment by mikegee** [comment]
Ruby hoists variables such that `animal` is `nil` even before entering the `if` statement.

http://stackoverflow.com/a/12928261/55878

**Comment by jonas054** [comment]
Look at is this way. If you change the name of the block parameter to `animal2` for example, you can reference both `animal` (which is `nil` and admittedly not much use, but still a defined variable) and `animal2` inside the block. But in your example's current state you can only reference the inner `animal` because it shadows the outer one.
